optaplanner-examples/src/main/java/org/optaplanner/examples/rocktour/optional/score/RockTourConstraintProvider.java

```diff
@@ -0,0 +1,82 @@
+import static org.optaplanner.examples.rocktour.domain.RockTourConstraintConfiguration.SHORTEN_DRIVING_TIME_PER_MILLISECOND_SQUARED;
+import static org.optaplanner.examples.rocktour.domain.RockTourConstraintConfiguration.UNASSIGNED_SHOW;
+
+public class RockTourConstraintProvider implements ConstraintProvider {
+
+    @Override
+    public Constraint[] defineConstraints(ConstraintFactory constraintFactory) {
+        return new Constraint[] {
+                requiredShow(constraintFactory),
+                unassignedShow(constraintFactory),
+                revenueOpportunity(constraintFactory),
+                drivingTimeToShowPerSecond(constraintFactory),
+                drivingTimeToBusArrivalPerSecond(constraintFactory),
+                delayShowCostPerDay(constraintFactory),
+                shortenDrivingTimePerMillisecondSquared(constraintFactory)
+        };
+    }
+
+    private UniConstraintStream<RockShow> getShowWithoutDate(ConstraintFactory constraintFactory) {
+        return constraintFactory.fromUnfiltered(RockShow.class)
+                .filter(r -> r.getDate() == null);
+    }
+
+    private Constraint requiredShow(ConstraintFactory constraintFactory) {
+        return getShowWithoutDate(constraintFactory)
+                .filter(RockShow::isRequired)
+                .penalizeConfigurable(REQUIRED_SHOW);
+    }
+
+    private Constraint unassignedShow(ConstraintFactory constraintFactory) {
+        return getShowWithoutDate(constraintFactory)
+                .filter(r -> r.getBus() != null)
+                .penalizeConfigurable(UNASSIGNED_SHOW);
+    }
+
+    private Constraint revenueOpportunity(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(RockShow.class)
+                .filter(r -> r.getDate() != null)
+                .rewardConfigurableLong(REVENUE_OPPORTUNITY, RockShow::getRevenueOpportunity);
+    }
+
+    private Constraint delayShowCostPerDay(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(RockShow.class)
+                .filter(r -> r.getDate() != null)
+                .penalizeConfigurableLong(DELAY_SHOW_COST_PER_DAY, RockShow::getDaysAfterBusDeparture);
+    }
+
+    private Constraint drivingTimeToShowPerSecond(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(RockShow.class)
+                .filter(r -> r.getPreviousStandstill() != null)
+                .penalizeConfigurableLong(DRIVING_TIME_TO_SHOW_PER_SECOND,
+                        RockShow::getDrivingTimeFromPreviousStandstill);
+    }
+
+    private Constraint shortenDrivingTimePerMillisecondSquared(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(RockShow.class)
+                .filter(r -> r.getPreviousStandstill() != null)
+                .penalizeConfigurableLong(SHORTEN_DRIVING_TIME_PER_MILLISECOND_SQUARED,
+                        r -> (long) Math.pow(r.getDrivingTimeFromPreviousStandstill(), 2));
```

Comment:
Math.pow does long to double conversation, which is inaccurate, try (double) 9007199254740993L
  https://ge0ffrey.github.io/ge0ffrey-presentations/cornerCaseCheatSheet/cheatSheetJava.html
then just to convert it back to long (again, inaccurate)

Lets simply do r.getDrivingTimeFromPreviousStandstill() * r.getDrivingTimeFromPreviousStandstill() ?

Reply:
This sounds more like an argument to avoid using `double`. What do we do when we want to ^50? We write a for loop?

Anyway... changing it.

Reply:
thanks, in any case - at some point, we will write penalizeUnfairness or .rewardLoadBalance or something like that, replacing this code, making this discussion mute